### PR TITLE
Mtp sampling

### PR DESCRIPTION
**Pull request by mahmoudhas**
# PR title

Please write the PR title by following template:

[JIRA ticket link/nvbug link/github issue link][fix/feat/doc/infra/...] \<summary of this PR\>

For example, assume I have a PR hope to support a new feature about cache manager of Jira TRTLLM-1000 ticket, it would be like

[TRTLLM-1000][feat] Support a new feature about cache manager

## Description

Please explain the issue and the solution in short.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/speculative/mtp.py [hunk 1]
```diff
@@ -237,6 +244,120 @@ def update_requests(self, decoder_state: DecoderState) -> None:
...
+            is_custom = False
+
+            if request.guided_decoding_params is not None:
+                guided_requests.append(i)
+                guided_logits_idx.append(cur_idx)
+                continue
+
+            if (request.sampling_config.temperature is not None
+                and request.sampling_config.temperature[0] > 0
+                and not math.isclose(request.sampling_config.temperature[0], self.DEFAULT_TEMPERATURE)):
```
Comment (bdubayah) [inline tensorrt_llm/_torch/speculative/mtp.py:287]: what's the point of the check against default temperature?

File: tensorrt_llm/_torch/speculative/mtp.py [hunk 2]
```diff
@@ -237,6 +244,120 @@ def update_requests(self, decoder_state: DecoderState) -> None:
...
+            cur_idx = next_idx
+            next_idx += 1 + request.num_draft_tokens
+            is_custom = False
+
+            if request.guided_decoding_params is not None:
+                guided_requests.append(i)
+                guided_logits_idx.append(cur_idx)
+                continue
+
+            if (request.sampling_config.temperature is not None
```
Comment (bdubayah) [inline tensorrt_llm/_torch/speculative/mtp.py:285]: For temperature == 0, need to add a special case that maps top_k to 1 (equivalent to greedy). unless greedy requests are handled somewhere differently i'm not seeing

File: tensorrt_llm/_torch/speculative/mtp.py [hunk 3]
```diff
@@ -237,6 +245,111 @@ def update_requests(self, decoder_state: DecoderState) -> None:
...
+        lens = model_outputs['new_tokens_lens']
+
+        sampled_token_idx = lens[request_idx] - 1
+        idxes = logits_idx + sampled_token_idx
+
+        logits = logits[idxes,:]
+        if len(temperature_idx) > 0:
+            logits[temperature_idx,:] /= torch.tensor(temperature, device=logits.device, dtype=logits.dtype).unsqueeze(1)
+
+        sampled_tokens = flashinfer.top_k_top_p_sampling_from_logits(logits, self.DEFAULT_TOP_K, top_p)
```
Comment (bdubayah) [inline tensorrt_llm/_torch/speculative/mtp.py:342]: it seems like the logits that are added to self.logits_idx are the logits for the first token generated by the base model:
```
cur_idx = next_idx
next_idx += 1 + request.num_draft_tokens
```
don't we want to be sampling on the last accepted draft token, since those draft tokens were already accepted using the greedily sampled logits? but here it seems like we only sample the first logit of each request. Maybe i am missing something

Comment (mahmoudhas) [reply tensorrt_llm/_torch/speculative/mtp.py:342]: I add the accepted length minus one to these before using them, so I’m actually using the logits of the last accepted token

File: tensorrt_llm/_torch/speculative/mtp.py (hunk 1 above)
Comment (mahmoudhas) [reply tensorrt_llm/_torch/speculative/mtp.py:297]: to know if the user has requested or not. if they haven't requested either temperature or top_p, we skip flashinfer sampling andj ust use the MTP-accepted token

File: tensorrt_llm/_torch/speculative/mtp.py (hunk 2 above)
Comment (mahmoudhas) [reply tensorrt_llm/_torch/speculative/mtp.py:295]: great catch! I'll treat temperature==0 requests as guided requests, effectively applying argmax for sampling

File: tensorrt_llm/_torch/speculative/mtp.py (hunk 1 above)
Comment (mahmoudhas) [reply tensorrt_llm/_torch/speculative/mtp.py]: My understanding is that temp=1.0 and top_p=1.0 are the default values. So why treat sampling in a special way?

https://platform.openai.com/docs/api-reference/responses/create#responses-create-temperature
https://platform.openai.com/docs/api-reference/responses/create#responses-create-top_p

both of them default to 1.0

File: tensorrt_llm/_torch/speculative/mtp.py (hunk 2 above)
Comment (mahmoudhas) [reply tensorrt_llm/_torch/speculative/mtp.py]: done with the refactor

File: tensorrt_llm/_torch/pyexecutor/decoder.py [hunk 4]
```diff
@@ -176,6 +178,8 @@ def __init__(self, max_seq_len: int, mixed_decoder: bool = False):
         self.max_seq_len = max_seq_len
         self.mixed_decoder = mixed_decoder
 
+        torch.manual_seed(42)
```
Comment (bdubayah) [inline tensorrt_llm/_torch/pyexecutor/decoder.py:181]: Is seeding it once enough to guarantee consistency across ranks?

File: tensorrt_llm/_torch/pyexecutor/b10.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+                is_custom = True
+                if temperature < 1e-6:
+                    is_greedy = True
+                    temperature = B10Decoder.DEFAULT_TEMPERATURE
+
+            if (top_p is not None
+                and top_p > 0
+                and top_p < 1):
+                assert len(sampling_config.top_p) == 1
+                is_custom = True
```
Comment (bdubayah) [inline tensorrt_llm/_torch/pyexecutor/b10.py:78]: So all requests will have is_custom = True, since top defaults to 1.0

**Review by bdubayah — APPROVED** [review]
LGTM. thanks for adding this to the regular path!
